### DISCUSSION TITLE
Add dynamic canonical links

### DISCUSSION THREAD

**Discussion by rafael-franca** [remix-run/remix#4867] [Q&A] [answered]
Hey, folks!

I want to add dynamic canonicals, but the links function not receive data from loaders.

How I can make this?

**Comment by sergiodxa** [comment, marked as answer]
You can use [Remix Utils's DynamicLinks](https://github.com/sergiodxa/remix-utils#dynamiclinks), but AFAIK the canonical link tag is body safe so you could render it in the body itself.

**Reply by rafael-franca** [reply]
Thanks for response. I wasn't know Remix Utils, but reading about this I notice that resolve some problems what I can have. Congratulations for idea of create this wonderful package.

**Reply by aaronadamsCA** [reply]
According to Google, [the canonical link must appear in `<head>`](https://developers.google.com/search/docs/crawling-indexing/consolidate-duplicate-urls#rel-canonical-link-method:~:text=The%20rel%3D%22canonical%22%20link%20element%20is%20only%20accepted%20if%20it%20appears%20in%20the%20%3Chead%3E%20section%20of%20the%20HTML%2C%20so%20make%20sure%20at%20least%20the%20%3Chead%3E%20section%20is%20valid%20HTML.), not `<body>`:

> The `rel="canonical"` link element is only accepted if it appears in the `<head>` section of the HTML, so make sure at least the [`<head>` section is valid HTML](https://developers.google.com/search/docs/crawling-indexing/valid-page-metadata).

**Comment by phuoccss** [comment]
I'm doing sth along this line. Maybe it helps

```js
// root.jsx
export function loader({ request, context }) {
  return json({
    origin: new URL(request.url).origin,
  })
}

export default RootRoute () {
  const { origin } = useLoaderData()
  const { pathname } = useLocation()
  const href = origin + pathname
  
  return (
    <html lang='en'>
      <head>
        ...
        <link href={href} rel='canonical' />
        ...
      </head>
      ...
    </html>
  )
}
```

If you have canonical link that is different from the page url, then would need to use the Remix's `handle` export.

**Reply by rafael-franca** [reply]
Thanks for the response! This works for me in some cases, but not when pathname is different of canonical.

**Reply by thomasverleye** [reply]
Hi @rafael-franca when is the pathname different from the canonical? Except maybe trailing slash?
